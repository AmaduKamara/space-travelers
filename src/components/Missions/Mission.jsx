import React from 'react';
import PropTypes from 'prop-types';

const Mission = ({
  id, description, missionName, toggleMission, reserved,
}) => (
  <>
    <tbody className="odd:bg-white even:bg-slate-200 border-2 border-slate-100">
      <tr>
        <td className="px-6 py-4  whitespace-wrap flex align-top text-sm font-bold text-gray-900">{missionName}</td>
        <td className="text-sm border-2 border-slate-100 text-gray-900 tracking-tight font-medium px-6 py-4 ">
          {description}
        </td>
        <td className="text-sm border-2 border-slate-100 text-gray-900 font-light px-6 py-4 whitespace-nowrap">
          {reserved ? <span className="text-xs px-2 py-0.5 font-bold bg-blue-500 text-white rounded">Active Member</span> : <span className="text-xs px-2 py-0.5 font-bold bg-gray-600 text-white rounded">NOT A MEMBER</span>}
        </td>
        <td className="text-sm text-gray-900 font-light px-6 py-4 whitespace-nowrap">
          {reserved ? <button type="button" id={id} onClick={toggleMission} className="bg-transparent hover:border-red-900 text-red-400 font-bold py-2 px-4 border border-red-400 rounded">Leave Mission</button> : <button type="button" id={id} onClick={toggleMission} className="bg-transparent hover:border-black font-bold py-2 px-4 border border-gray-400 text-black rounded">Join Mission</button>}

        </td>
      </tr>
    </tbody>
  </>
);

export default Mission;

Mission.propTypes = {
  id: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  missionName: PropTypes.string.isRequired,
  toggleMission: PropTypes.func.isRequired,
  reserved: PropTypes.bool,

};

Mission.defaultProps = {
  reserved: false,
};
