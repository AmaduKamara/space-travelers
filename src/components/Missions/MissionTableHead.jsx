import React from 'react';

const MissionTableHead = () => (
  <>
    <thead className="bg-white border-b">
      <tr>
        <th scope="col" className="text-sm font-bold text-gray-900 px-6 py-4 text-left border">
          Mission
        </th>
        <th scope="col" className="text-sm font-bold text-gray-900 px-6 py-4 text-left border">
          Description
        </th>
        <th scope="col" className="text-sm font-bold text-gray-900 px-6 py-4 text-left border">
          Status
        </th>
        <th scope="col" className="text-sm font-bold text-gray-900 px-6 py-4 text-left">
          {' '}
        </th>
      </tr>
    </thead>
  </>
);

export default MissionTableHead;
