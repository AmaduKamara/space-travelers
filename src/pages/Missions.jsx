import React from 'react';

function Mission() {
  return (
    <div className="container mx-auto">
      <h1>Mission Page</h1>
    </div>
  );
}

export default Mission;
