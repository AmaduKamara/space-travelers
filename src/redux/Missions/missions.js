// CONSTANTS
const INITIAL_STATE = {
  missions: [],
};
const FETCH_MISSIONS = 'Missions/FETCH_MISSIONS';
const TOGGLE_MISSION = 'Missions/TOGGLE_MISSION';

// Local ACTIONS
export const fetchMissions = (payload) => ({
  type: FETCH_MISSIONS,
  payload,
});

export const joinMission = (payload) => ({
  type: TOGGLE_MISSION,
  payload,
});

export const toggleMission = (state, payload) => {
  const newState = state.map((missions) => {
    if (missions.mission_id !== payload) return missions;
    return { ...missions, reserved: !missions.reserved };
  });
  return newState;
};

// REDUCER
const reducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case FETCH_MISSIONS:
      return { ...state, missions: action.payload };
    case TOGGLE_MISSION:
      return { missions: toggleMission(state.missions, action.payload) };
    default:
      return state;
  }
};

export default reducer;
