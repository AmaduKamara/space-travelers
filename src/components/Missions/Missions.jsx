import { React, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions } from '../../redux/Missions/missions';
import Mission from './Mission';
import spinner from '../../Images/spinner.gif';

const Missions = () => {
  const dispatch = useDispatch();
  const selector = useSelector((state) => state.missionsData);
  useEffect(() => {
    dispatch(fetchMissions());
  }, []);

  return (
    <>
      <div className="flex flex-col">
        <div className="overflow-x-auto sm:-mx-6 lg:-mx-8">
          <div className="py-2 inline-block min-w-full sm:px-6 lg:px-8">
            <div className="overflow-hidden">
              <table className="min-w-full border">
                <thead className="bg-white border-b">
                  <tr>
                    <th scope="col" className="text-sm font-medium text-gray-900 px-6 py-4 text-left">
                      Mission
                    </th>
                    <th scope="col" className="text-sm font-medium text-gray-900 px-6 py-4 text-left">
                      Description
                    </th>
                    <th scope="col" className="text-sm font-medium text-gray-900 px-6 py-4 text-left">
                      Status
                    </th>
                    <th scope="col" className="text-sm font-medium text-gray-900 px-6 py-4 text-left">
                      {' '}
                    </th>
                  </tr>
                </thead>
                {selector.isLoading ? (
                  <span>
                    <img src={spinner} alt="spinner" />
                    {' '}
                    Loading ...
                  </span>
                ) : (selector.missions.map((items) => {
                  const { mission_id: missionID, mission_name: missionName, description } = items;
                  return (
                    <Mission
                      key={missionID}
                      id={missionID}
                      description={description}
                      missionName={missionName}
                    />
                  );
                })) }
              </table>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default Missions;
