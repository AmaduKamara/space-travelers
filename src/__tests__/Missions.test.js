import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import Missions from '../components/Missions/Missions';
import store from '../redux/configStore';

const renderer = require('react-test-renderer');

it('matches snapshot of Missons component', () => {
  const tree = renderer.create(<Provider store={store}><Missions /></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders table headers', () => {
  render(<Provider store={store}><Missions /></Provider>);
  const element = screen.getByText(/Mission/i);
  expect(element).toBeInTheDocument();
});
