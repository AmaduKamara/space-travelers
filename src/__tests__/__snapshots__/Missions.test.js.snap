// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot of Missons component 1`] = `
<div
  className="flex flex-col mb-16"
>
  <div
    className="overflow-x-auto sm:-mx-6 lg:-mx-8"
  >
    <div
      className="py-2 inline-block min-w-full sm:px-6 lg:px-8"
    >
      <div
        className="overflow-hidden"
      >
        <table
          className="min-w-full border"
        >
          <thead
            className="bg-white border-b"
          >
            <tr>
              <th
                className="text-sm font-bold text-gray-900 px-6 py-4 text-left border"
                scope="col"
              >
                Mission
              </th>
              <th
                className="text-sm font-bold text-gray-900 px-6 py-4 text-left border"
                scope="col"
              >
                Description
              </th>
              <th
                className="text-sm font-bold text-gray-900 px-6 py-4 text-left border"
                scope="col"
              >
                Status
              </th>
              <th
                className="text-sm font-bold text-gray-900 px-6 py-4 text-left"
                scope="col"
              >
                 
              </th>
            </tr>
          </thead>
        </table>
      </div>
    </div>
  </div>
</div>
`;
